safety-architecture/tools/callgraph-tool/doc/find_callchains_examples.md

```diff
@@ -0,0 +1,98 @@
+  --calls CALLS         function call database csv file
+  --from_function FROM_FUNCTION
+                        function name from where the search begins (literal
+                        match)
+  --to_function TO_FUNCTION
+                        function name where call chain ends (regex match)
+
+Example: ./find_callchains.py --from_function sock_recvmsg --to_function
+'^x64_sys'--calls calls.csv
+```
+
+
+The script exports the results of a user search either to a `CSV` file (if `out` input parameter ends with .csv) or to an image file otherwise. The source function is always a literal match, while the destination function can be defined as regular expression. The search can be performed in both directions, i.e. both in
+callers of the source function - `left` search and callees of the source function - `right` search.
+
+# Usage examples
+
+
+Suppose that we have an example database for a program with a following call graph:
+
+<img src="all_callchains.jpg">
+<br /><br />
+
+## Search for callers
+
+Say we are interested if the `say_hello` function is in callee list of `start_of_longer_call_chain` function. We can issue the following call to generate the path between those two functions if it exists:
+```
+./find_callchains.py --from_function say_hello --to_function start_of_longer_call_chain --direction left --out long_chain.jpg --calls calls.csv
+```
+
+The result of the query is in the following image:
+
+<img src="long_chain.jpg">
+<br /><br />
+
+## Search for callees
+
+We can also search in forward direction to find out the paths from a caller `main` to a specific function `say_hello` we are interested in:
+```
+./find_callchains.py --from_function main --to_function say_hello --direction right --out multi_chain.jpg --calls calls.csv
+```
+The command output is shown in:
+
+<img src="multi_chain.jpg">
+<br /><br />
+
+## Search for both callers and callees
+
+If we want to find out if any of the `chainX` functions, where `X` is a single-digit number, is linked either as a caller or callee to `chain2` function we can issue the following query:
+```
+./find_callchains.py --from_function chain2 --to_function chain[0-9] --direction both --out chainX.jpg --calls calls.csv
+```
+This query results in a following image for a given database:
+
+<img src="chainX.jpg">
+<br /><br />
+
+## Larger example
+The examples given so far are intentionally contrived to illustrate the capabilites of the script. These results could have easily been read out from the image of the whole program graph. The usefulness of the script can be demonstrated on a more complicated codebase, such as that of a Linux kernel. Say we want to find out the receive system calls that use `sock_recvmsg` function. We can answer this question by issuing the following query:
+```
+./find_callshains --calls calls.csv, --from_function sock_recvmsg, --to_function ^__x64_sys_recv, --direction=left, --cutoff=12, --out=from_sock_recvmsg_to_x64_sys_recv_left.jpg
```

Comment:
Are the commas in this command intentional?

Reply:
They remained as I copied command from IDE. Will be fixed